### PR TITLE
Feature/vram

### PR DESCRIPTION
**Pull request by dolovnyak**

### CODE REVIEW

File: include/device/GB_vram.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <common/GB_macro.h>
+#include "GB_config.h"
+
+namespace GB::device {
+
+class VRAM {
+ public:
+    struct registers_t {
+        constexpr static u32 RESERVED_BITS = 0xFE;  // TODO(hgranule) I didn't understand how to set range of
```
Comment (hgranule) [inline include/device/GB_vram.h:12]:
```cpp
::bits_set(0, 2, 5, 3, 1);  // 0b00101111
::bit_mask(5, 1); // 0b00111110
```

File: include/device/GB_vram.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#define DEVICE_GB_VRAM_H_
+
+#include <common/GB_macro.h>
+#include "GB_config.h"
+
+namespace GB::device {
+
+class VRAM {
+ public:
+    struct registers_t {
```
Comment (hgranule) [inline include/device/GB_vram.h:11]: Registers

standard-like types must be `${type_name}_t`;
All other-types must be -> CamelCased

File: include/device/GB_vram.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    dbuffer_t   __memory;
+    registers_t __regs;
+
+    inline unsigned __calc_phys_addr(word_t inner_vaddr) const;
+};
+
+/**
+ * @brief   Return VBK register value (only zero bit readable).
+ *
+ * @details In DMG mode it always return 0xFE (default VBK_INIT_VALUE), zero bit always set in '0', other always set in '1'.
```
Comment (hgranule) [inline include/device/GB_vram.h:47]: let it return 0xFE right now, but mark it with
```cpp
// TODO(hgranule): DMG/CGB shit
// <TABS> description
```

File: include/device/GB_vram.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+/**
+ * @brief   Write value to VBK register (only zero bit writable).
+ *
+ * @details In DMG mode it always does nothing.
+ *          In CGB mode only zero bit will rewrite.
+ */
+inline void
+VRAM::set_VBK_reg(byte_t value) {
```
Comment (hgranule) [inline include/device/GB_vram.h:62]: Delete switch/case, you can just NOT map this function (register) to memory!

File: include/device/GB_vram.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    GBModeFlag TEMPORARY_MODE_FLAG = CGB_MODE;  // TODO(dolovnyak) change to correct mode flag when it will be implemented
+
+    switch (TEMPORARY_MODE_FLAG) {
+        case DMG_MODE:
+            return;
+        case CGB_MODE:
+            __regs.VBK = value | registers_t::RESERVED_BITS;
```
Comment (hgranule) [inline include/device/GB_vram.h:69]: __regs.VBK = ::bit_n(0, value);

File: include/device/GB_vram.h [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * @brief   Return VBK register value (only zero bit readable).
+ *
+ * @details In DMG mode it always return 0xFE (default VBK_INIT_VALUE), zero bit always set in '0', other always set in '1'.
+ *          In CGB mode only zero bit matter (return current state), other seven always set in '1'.
+ */
+inline byte_t
+VRAM::get_VBK_reg() const {
+    return __regs.VBK;
```
Comment (hgranule) [inline include/device/GB_vram.h:52]: return __regs.VBK | RESERVED_BITS;

File: include/device/GB_vram.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+inline void
+VRAM::write_phys_addr(word_t phys_addr, byte_t value) {
+    __memory[phys_addr] = value;
+}
+
+inline unsigned
+VRAM::__calc_phys_addr(word_t inner_vaddr) const {
+    const u32 bank_offset = inner_vaddr % VRAM_BANK_SIZE;
+    if (!(__regs.VBK & 1u))
```
Comment (hgranule) [inline include/device/GB_vram.h:100]: ::bit_n() or just raw value

Comment (hgranule) [reply include/device/GB_vram.h:100]:
```cpp
const u32 bank_base = (::bit_n(0, __regs.VBK)) ? VRAM_BANK_SIZE : 0;
return bank_base + bank_offset;
```

File: include/device/GB_wram.h [hunk 8]
```diff
@@ -18,7 +18,7 @@ namespace GB::device {
 
 class WRAM {
  public:
-    struct Registers {
+    struct Registers {  // TODO(hgranule) codestyle
```
Comment (hgranule) [inline include/device/GB_wram.h:21]: nope

File: include/memory/GB_vaddr.h [hunk 9]
```diff
@@ -113,8 +113,10 @@ enum VirtualAddress : word_t {
     /* OAM RAM */
     OAM_RAM_BASE_VADDR = 0xFE00,
     OAM_RAM_LAST_VADDR = 0xFE9F,
-    OAM_ECHO_BASE_VADDR = 0xFEA0,  // NOTE: There is article with behavior for UNUSED_AFTER_OAM memory
-    OAM_ECHO_LAST_VADDR = 0xFEFF,  //       https://github.com/AntonioND/giibiiadvance/blob/master/docs/TCAGBD.pdf (2.10)
+    OAM_ECHO_BASE_VADDR = 0xFEA0,   // NOTE: There is article with behavior for UNUSED_AFTER_OAM memory
+    OAM_ECHO_LAST_VADDR = 0xFEFF,   //       https://github.com/AntonioND/giibiiadvance/blob/master/docs/TCAGBD.pdf (2.10)
+                                    // TODO(dolovnyak, hgranule): I want to create device GB_echo_oram int the future,
```
Comment (hgranule) [inline include/memory/GB_vaddr.h:118]: in*

File: include/memory/GB_vaddr.h [hunk 10]
```diff
@@ -113,8 +113,10 @@ enum VirtualAddress : word_t {
...
+                                    //                            for implementation correct read/write behavior
```
Comment (hgranule) [inline include/memory/GB_vaddr.h:119]: implementation of shit
implement shit

File: test/vram.cc [hunk 11]
```diff
@@ -0,0 +1,107 @@
+#include "gtest/gtest.h"
+
+#include "device/GB_vram.h"
+
+namespace {
+
+using VRAM = GB::device::VRAM;
+
+class VRAM_mock : public VRAM {
```
Comment (hgranule) [inline test/vram.cc:9]: TODO(dolovnyak):
After last infrastructure change remove MOck and use original class with redefined protected

File: test/vram.cc [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    vram_mock.write_phys_addr(0x100, 101);
+    vram_mock.write_phys_addr(0x1FFF, 102);
+    vram_mock.write_phys_addr(0x3FFF, 103);
+
+    EXPECT_EQ(vram_mock.read_phys_addr(0x0), 100);
+    EXPECT_EQ(vram_mock.read_phys_addr(0x100), 101);
+    EXPECT_EQ(vram_mock.read_phys_addr(0x1FFF), 102);
+    EXPECT_EQ(vram_mock.read_phys_addr(0x3FFF), 103);
+
+    EXPECT_EQ(vram_mock.get_VBK_reg(), 0xFE);
```
Comment (hgranule) [inline test/vram.cc:45]: Split this test to the phys RW and the inner RW

File: test/vram.cc [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+    EXPECT_EQ(vram_mock.get_memory_ref().size(), 16_KBytes);
+
+    VRAM vram;
+
+    vram.write_phys_addr(0, 100);
+    vram.write_phys_addr(10, 101);
+    vram.write_phys_addr(159, 102);
+
+    VRAM vram_test_move_constr(std::move(vram));
```
Comment (hgranule) [inline test/vram.cc:82]: Please check if dbuffer of vram was moved to that one
Also do it for ORAM

File: test/wram.cc [hunk 14]
```diff
@@ -1,4 +1,4 @@
-#include <utility>
+#include <utility>  // TODO(hgranule) it not used
```
Comment (hgranule) [inline test/wram.cc:1]: Delete it

Comment (hgranule) [reply test/wram.cc:1]: NOT UR COMMNSDA
DELETE MY HEADER DECLARATION

File: include/GB_config.h [hunk 15]
```diff
@@ -10,20 +10,27 @@
...
-    constexpr unsigned WRAM_CGB_BANK_SIZE = 4_KBytes;
-    constexpr unsigned WRAM_CGB_SIZE = 32_KBytes;
-    constexpr unsigned WRAM_NON_CGB_SIZE = 8_KBytes;
+constexpr unsigned WRAM_CGB_BANK_SIZE = 4_KBytes;
+constexpr unsigned WRAM_CGB_SIZE = 32_KBytes;
+constexpr unsigned WRAM_NON_CGB_SIZE = 8_KBytes;
 
-    constexpr unsigned SVBK_INIT_VALUE = 0x0;
+constexpr unsigned SVBK_INIT_VALUE = 0x0;
+constexpr unsigned VBK_INIT_VALUE = 0xFE;
```
Comment (hgranule) [inline include/GB_config.h:18]: 0x0

File: include/GB_config.h [hunk 16]
```diff
@@ -10,20 +10,27 @@
...
+constexpr unsigned VRAM_BANK_SIZE = 8_KBytes;
+constexpr unsigned VRAM_NON_CGB_SIZE = VRAM_BANK_SIZE;
+constexpr unsigned VRAM_CGB_SIZE = VRAM_BANK_SIZE * 2;
+constexpr unsigned VRAM_MAX_SIZE = VRAM_CGB_SIZE;
+
+
+enum GBModeFlag : u16 {
+    DMG_MODE = 0b000001,
+    MGB_MODE = 0b000010,
+    CGB_MODE = 0b001000,
```
Comment (hgranule) [inline include/GB_config.h:32]: can you add plkease CGB_DMG_MODE = 0b100

File: include/common/GB_types.h [hunk 17]
```diff
@@ -25,7 +25,7 @@ using i8    = int8_t;
 using word_t  = u16;  ///< 16-bit word type
 using byte_t  = u8;   ///< 8-bit byte type
 
-using Reg8 = byte_t;
+using Reg8 = byte_t;  // TODO(hgranule) codesyle
```
Comment (hgranule) [inline include/common/GB_types.h:28]: Delete it

File: include/device/GB_vram.h [hunk 18]
```diff
@@ -0,0 +1,108 @@
+#ifndef DEVICE_GB_VRAM_H_
+#define DEVICE_GB_VRAM_H_
+
+#include <common/GB_macro.h>
```
Comment (hgranule) [inline include/device/GB_vram.h:4]:
```cpp
#include "OUR HEADERS"
```

File: test/wram.cc [hunk 19]
```diff
@@ -1,4 +1,4 @@
...
 #include <algorithm>
```
Comment (hgranule) [inline test/wram.cc:2]: Delete it too

**Review by hgranule — CHANGES_REQUESTED** [review]

**Review by hgranule — APPROVED** [review]
Looks good

**Comment by hgranule** [comment]
For the future:
I think, it's better to commit your "review fixes" commit separately from the previous reviewed commit. That makes review simpler. Let's use `git commit --amend` when there are only 1-3 little fixes, or we can just not use it anymore and use SQUASH and MERGE. What do you think about that???

UPD: Amending and squashing for the master branch. When you work on your Merge Request, better to commit everything separately.

**Comment by dolovnyak** [comment]
> For the future:
> I think, it's better to commit your "review fixes" commit separately from the previous reviewed commit. That makes review simpler. Let's use `git commit --amend` when there are only 1-3 little fixes, or we can just not use it anymore and use SQUASH and MERGE. What do you think about that???
> 
> UPD: Amending and squashing for the master branch. When you work on your Merge Request, better to commit everything separately.

Ok

**Pull request merged**